extensions/src/main/java/io/lionweb/repoclient/RepoSerialization.java

```diff
@@ -0,0 +1,261 @@
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+import kotlin.text.Charsets;
+
+/** This class contains the logic to store and retrieve entire repositories at once. */
+public class RepoSerialization {
+  private int nNodesThreshold = 100_000;
+  private TransferFormat transferFormat = TransferFormat.FLATBUFFERS;
+  private Compression compression = Compression.DISABLED;
+
+  /**
+   * Download all the content of the repository accessed by the apiClient into a directory. In this
+   * directory one file per partition is created. The file is in JSON format.
+   */
+  public <C extends RawBulkAPIClient & BulkAPIClient> void downloadRepoAsDirectory(
+      C apiClient, File directory) throws IOException {
+    for (String partitionID : apiClient.listPartitionsIDs()) {
+      String partitionData =
+          apiClient.rawRetrieve(Collections.singletonList(partitionID), Integer.MAX_VALUE);
+      File partitionFile = new File(directory, partitionID + ".json");
+      Files.write(
+          partitionFile.toPath(),
+          Collections.singletonList(partitionData),
+          StandardOpenOption.CREATE,
+          StandardOpenOption.TRUNCATE_EXISTING);
+    }
+  }
+
+  /**
+   * Download all the content of the repository accessed by the apiClient into a zip file. In this
+   * zip file, one entry per partition is created. The entry is in JSON format.
+   */
+  public <C extends RawBulkAPIClient & BulkAPIClient> void downloadRepoAsZip(
+      C apiClient, File zipFile) throws IOException {
+    try (FileOutputStream fos = new FileOutputStream(zipFile);
+        ZipOutputStream zos = new ZipOutputStream(fos)) {
+
+      for (String partitionID : apiClient.listPartitionsIDs()) {
+        String partitionData =
+            apiClient.rawRetrieve(Collections.singletonList(partitionID), Integer.MAX_VALUE);
+
+        ZipEntry entry = new ZipEntry(partitionID + ".json");
+        zos.putNextEntry(entry);
+
+        byte[] dataBytes = partitionData.getBytes(StandardCharsets.UTF_8);
+        zos.write(dataBytes, 0, dataBytes.length);
+
+        zos.closeEntry();
+      }
+    }
+  }
+
+  /**
+   * Upload all the content of a directory to a given repository, using the standard bulk API (and
+   * not the more performant bulk import). The directory and all the subdirectories are examined,
+   * looking for files with extension ".json" (ignoring case).
+   */
+  public void simpleUploadDirectoryToRepo(RawBulkAPIClient apiClient, File directory)
+      throws IOException {
+    if (!directory.isDirectory()) {
+      throw new IllegalArgumentException(
+          "Provided file is not a directory: " + directory.getAbsolutePath());
+    }
+
+    for (File file : findJsonFilesRecursively(directory)) {
+      String content = new String(Files.readAllBytes(file.toPath()), Charsets.UTF_8);
```

Comment:
This means your whole file needs to fit in memory twice -- once as string, once as SerializedNodes.

Reply:
Sadly yes. Unfortunately this is the case because we currently cannot create partitions with children: we must create first without children and then add the children in a second call. So we need to process the string into a data structure that we can manipulate to remove children.